diff --git a/CMakeLists.txt b/CMakeLists.txt
index b65f1b5..19cc5a9 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -62,12 +62,13 @@ include(OCL)
 set(AMDGCN_LIB_LIST)
 set(AMDGCN_DEP_LIST)
 add_subdirectory(irif)
-add_subdirectory(oclc)
-add_subdirectory(ocml)
-add_subdirectory(ockl)
-add_subdirectory(opencl)
-add_subdirectory(hip)
-add_subdirectory(asanrtl)
+#add_subdirectory(oclc)
+#add_subdirectory(ocml)
+#add_subdirectory(ockl)
+#add_subdirectory(opencl)
+#add_subdirectory(hip)
+#add_subdirectory(asanrtl)
+add_subdirectory(cuda2gcn)
 
 enable_testing()
 add_subdirectory(test/compile)
